sdks/java/io/cassandra/src/test/java/org/apache/beam/sdk/io/cassandra/CassandraIOTest.java

```diff
@@ -0,0 +1,231 @@
+                    return kv;
+                  }
+                }));
+    PAssert.that(mapped.apply("Count occurrences per scientist", Count.<String, Integer>perKey()))
+        .satisfies(
+            new SerializableFunction<Iterable<KV<String, Long>>, Void>() {
+              @Override
+              public Void apply(Iterable<KV<String, Long>> input) {
+                for (KV<String, Long> element : input) {
+                  assertEquals(element.getKey(), 1000, element.getValue().longValue());
+                }
+                return null;
+              }
+            });
+
+    pipeline.run();
+  }
+
+  /**
+   * A {@link CassandraService} implementation that stores the entity in memory.
+   */
+  private static class FakeCassandraService implements CassandraService {
+
+    private static final Map<Integer, Scientist> table = new HashMap<>();
+
+    public void load() {
+      table.clear();
+      String[] scientists = {
+          "Einstein",
+          "Darwin",
+          "Copernicus",
+          "Pasteur",
+          "Curie",
+          "Faraday",
+          "Newton",
+          "Bohr",
+          "Galilei",
+          "Maxwell"
+      };
+      for (int i = 0; i < 10000; i++) {
+        int index = i % scientists.length;
+        Scientist scientist = new Scientist();
+        scientist.id = i;
+        scientist.name = scientists[index];
+        put(scientist);
+      }
+    }
+
+    public void put(Scientist scientist) {
+      table.put(scientist.id, scientist);
+    }
+
+    public Scientist get(int id) {
+      return table.get(id);
+    }
+
+    @Override
+    public Reader createReader(String query, Class entity) throws IOException {
+      return new FakeCassandraReader();
+    }
+
+    static class FakeCassandraReader implements CassandraService.Reader {
+
+      private Iterator<Scientist> iterator;
+      private Scientist current;
+
+      @Override
+      public boolean start() throws IOException {
+        iterator = table.values().iterator();
+        return advance();
+      }
+
+      @Override
+      public boolean advance() throws IOException {
+        if (iterator.hasNext()) {
+          current = iterator.next();
+          return true;
+        }
+        return false;
+      }
+
+      @Override
+      public void close() {
+        iterator = null;
+        current = null;
+      }
+
+      @Override
+      public Scientist getCurrent() throws NoSuchElementException {
+        if (current == null) {
+          throw new NoSuchElementException();
+        }
+        return current;
+      }
+    }
+
+    @Override
+    public long getEstimatedSizeBytes() {
+      long size = 0L;
+      for (Scientist scientist : table.values()) {
+        size = size + scientist.toString().getBytes().length;
+      }
+      return size;
+    }
+
+    @Override
+    public List<BoundedSource<Scientist>> splitIntoBundles(CassandraIO.Read spec, long
```

Comment:
This method seems to replace the quite complicated logic used in the actual implementation, leaving that implementation untested.